include/libopencm3/nrf/common/i2c.h

```diff
@@ -97,21 +104,47 @@
 #define I2C_ERRORSRC_ANACK         (1 << 1)
 #define I2C_ERRORSRC_DNACK         (1 << 2)
 
-#define I2C_ENABLE_VALUE           (5)
+/** @addtogroup i2c_mode I2C peripheral mode 
+ * @{
+ */
+/** NRF51 legacy mode. 
+ * On NRF51, this is the only mode available.
+ * On NRF52, this mode does not support EasyDMA.
+ */
+#define I2C_MODE_LEGACY           (5)
+/**@}*/
+
+/** @addtogroup i2c_freq_const I2C frequency constants
+ * @{
+ */
 
+/** 100kHz */
 #define I2C_FREQUENCY_100K         (0x01980000)
+/** 250kHz */
 #define I2C_FREQUENCY_250K         (0x04000000)
+/** 390kHz
+ * @note: This value is not documented in datasheet. It provides
+ * ~390kHz clock with correct timing.
+ */
+#define I2C_FREQUENCY_390K         (0x06200000)
+/** 400kHz
+ * @note: According to datasheet, there is HW bug which prevents
+ * MCU from generating correct timings, therefore it might be
+ * unusable. Use @ref I2C_FREQUENCY_390K instead, if this affects
```

Comment:
lol :)

Reply:
Related to this, a) does this apply to nrf51 as well as nrf52? (it's in the common defines here) and b) should "set fast mode" just do a "set 390" behind the scenes on the affected targets?

Reply:
According to this (https://infocenter.nordicsemi.com/topic/errata_nRF52832_Rev3/ERR/nRF52832/Rev3/latest/anomaly_832_219.html), this bug is only present on NRF52 *and* only if I2C_MODE_MASTER is used. While the difference in timing is some 50ns , certain devices even might work, thus automatic setting of 390kHz instead of 400kHz is probably not what we want to do here.

The whole device is riddled with so many bugs, that I am not even sure what to do. Some of them can have workarounds implemented at level LOCM3 represents, but some workarounds are too heavy and complex to be implemented in LOCM3.

Reply:
sounds reasonable, leave it as you have it then, it all looks good other than that oddly worded doc just above :)